session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift

```diff
@@ -0,0 +1,231 @@
+    /// - "update" - if certain `id` exists in both arrays and `isDifferent(than:)` returns `true`.
+    var id: DiffableID { get }
+
+    /// Determines if this element has any changes if compared to other element.
+    /// It is used from `computeDiff(oldArray:newArray:)`. Both elements are guaranteed to have the same `id`.
+    /// - Parameter otherElement: the element to compare against
+    /// - Returns: `true` if data in both elements is different; `false` otherwise
+    func isDifferent(than otherElement: Self) -> Bool
+}
+
+/// Describes differences between both arrays compared in `computeDiff(oldArray:newArray:)`. It lists minimal
+/// set of changes that needs to be applied to `oldArray` to result with `newArray`.
+///
+/// Note:
+/// - when reconstructing `newArray` from `oldArray` and `Diff` **removals must be applied before additions**;
+/// - if certain element changes its position in `newArray`, `Diff` includes its **removal and addition** (on new position).
+internal struct Diff<Element: Diffable> {
+    /// Dictates that new element was added in `newArray`.
+    struct Add {
+        /// The `id` of preceding element in `newArray` or `nil` if new element should be inserted at index `0`.
+        let previousID: DiffableID?
+        /// The element to add.
+        let new: Element
+    }
+
+    /// Dictates that given element exists in both arrays but its data was changed.
+    struct Update {
+        /// Previous version of element (it has the same `id` as `to`).
+        let from: Element
+        /// New version of element (it has the same `id` as `from`).
+        let to: Element
+    }
+
+    /// Dictates that element was removed from `oldArray`.
+    struct Remove {
+        /// The `id` of element to remove.
+        let id: DiffableID
+    }
+
+    /// Elements to add.
+    let adds: [Add]
+    /// Elements to update (change data).
+    let updates: [Update]
+    /// Elements to remove.
+    let removes: [Remove]
+
+    /// If diff has no changes.
+    /// Empty diff indicates that arrays compared in `computeDiff(oldArray:newArray:)` are equal.
+    var isEmpty: Bool {
+        return adds.isEmpty && updates.isEmpty && removes.isEmpty
+    }
+}
+
+/// Thrown if unexpected happened in implemented algorithm.
+internal struct DiffError: Error {}
+
+/// An item in symbols table (`[DiffableID: Symbol]`) in Heckel's algorithm.
+///
+/// Note: Unlike original algorithm, our implementation doesn't use `OC`, `NC` and `OLNO` counters. In our case, elements within each array are unique,
+/// so instead of repetition counters, we define basic `Bool` flags to track occurence of certain `id` in one or both files.
+private struct Symbol {
+    /// If element with certain `id` occurs in `oldArray`.
+    var inOld: Bool
+    /// If element with certain `id` occurs in `newArray`.
+    var inNew: Bool
+    /// The index of element in `oldArray`.
+    var indexInOld: Int?
+}
```

Comment:
If `indexInOld` is nil it means that symbol is not in old. Then, this could be:
```swift
private struct Symbol {
    /// If element with certain `id` occurs in `newArray`.
    var inNew: Bool
    /// The index of element in `oldArray`.
    var indexInOld: Int?
}
```

It would simplify the condition below in `computeDiff`:
```swift
if symbol.inOld && symbol.inNew {
    guard let indexInOld = symbol.indexInOld else {
        throw DiffError()
    }
    na[index] = .index(indexInOld)
    oa[indexInOld] = .index(index)
}
```
to 
```swift
if symbol.inNew, let indexInOld = symbol.indexInOld {
    na[index] = .index(indexInOld)
    oa[indexInOld] = .index(index)
}
```

Reply:
Good idea, done 👌!